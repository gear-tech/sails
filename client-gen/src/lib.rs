mod generator;

use anyhow::{Context, Result};
use std::{fs, path::Path};

pub fn generate_client_from_idl(idl_path: &Path, out_path: &Path) -> Result<()> {
    let idl = fs::read_to_string(idl_path)
        .with_context(|| format!("Failed to open {} for reading", idl_path.display()))?;

    let program = match sails_idlparser::ast::parse_idl(&idl) {
        Ok(program) => program,
        Err(e) => {
            eprintln!("Failed to parse IDL: {}", e);
            std::process::exit(1);
        }
    };

    let buf = generator::generate(program).context("failed to generate client")?;

    fs::write(out_path, buf)
        .with_context(|| format!("Failed to write generated client to {}", out_path.display()))?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn full() {
        const IDL: &str = r#"
        type ThisThatSvcAppTupleStruct = struct {
            bool,
        };

        type ThisThatSvcAppDoThatParam = struct {
            p1: u32,
            p2: str,
            p3: ThisThatSvcAppManyVariants,
        };

        type ThisThatSvcAppManyVariants = enum {
            One,
            Two: u32,
            Three: opt u32,
            Four: struct { a: u32, b: opt u16 },
            Five: struct { str, u32 },
            Six: struct { u32 },
        };

        type T = enum { One };

        constructor {
            New : (a: u32);
        };

        service {
            DoThis : (p1: u32, p2: str, p3: struct { opt str, u8 }, p4: ThisThatSvcAppTupleStruct) -> struct { str, u32 };
            DoThat : (param: ThisThatSvcAppDoThatParam) -> result (struct { str, u32 }, struct { str });
            query This : (v1: vec u16) -> u32;
            query That : (v1: null) -> result (str, str);
        };
        "#;
        let program = sails_idlparser::ast::parse_idl(IDL).expect("parse IDL");

        let generated = generator::generate(program).unwrap();

        dbg!(&generated);

        insta::assert_snapshot!(generated);
    }
}
