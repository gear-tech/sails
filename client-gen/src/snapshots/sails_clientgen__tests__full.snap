---
source: client-gen/src/lib.rs
expression: generated
---
// Code generated by sails-client-gen. DO NOT EDIT.
use gstd::prelude::*;
use parity_scale_codec::{Decode, Encode};
use sails_sender::{Call, GStdSender};
pub trait Service {
    fn do_this(
        &mut self,
        p1: u32,
        p2: String,
        p3: (Option<String>, u8),
        p4: ThisThatSvcAppTupleStruct,
    ) -> Call<(String, u32)>;
    fn do_that(
        &mut self,
        param: ThisThatSvcAppDoThatParam,
    ) -> Call<Result<(String, u32), (String,)>>;
    fn this(&self, v1: Vec<u16>) -> Call<u32>;
    fn that(&self, v1: ()) -> Call<Result<String, String>>;
}

#[derive(Clone)]
pub struct Client<'a> {
    sender: &'a GStdSender,
}

impl<'a> Client<'a> {
    pub fn new(sender: &'a GStdSender) -> Self {
        Self { sender }
    }
}

impl<'a> Service for Client<'a> {
    fn do_this(
        &mut self,
        p1: u32,
        p2: String,
        p3: (Option<String>, u8),
        p4: ThisThatSvcAppTupleStruct,
    ) -> Call<(String, u32)> {
        Call::new(self.sender, "DoThis", (p1, p2, p3, p4))
    }
    fn do_that(
        &mut self,
        param: ThisThatSvcAppDoThatParam,
    ) -> Call<Result<(String, u32), (String,)>> {
        Call::new(self.sender, "DoThat", param)
    }
    fn this(&self, v1: Vec<u16>) -> Call<u32> {
        Call::new(self.sender, "This", v1)
    }
    fn that(&self, v1: ()) -> Call<Result<String, String>> {
        Call::new(self.sender, "That", v1)
    }
}
#[derive(PartialEq, Debug, Encode, Decode)]
pub struct ThisThatSvcAppTupleStruct(bool);
#[derive(PartialEq, Debug, Encode, Decode)]
pub struct ThisThatSvcAppDoThatParam {
    p1: u32,
    p2: String,
    p3: ThisThatSvcAppManyVariants,
}
#[derive(PartialEq, Debug, Encode, Decode)]
pub enum ThisThatSvcAppManyVariants {
    One,
    Two(u32),
    Three(Option<u32>),
    Four { a: u32, b: Option<u16> },
    Five((String, u32)),
    Six((u32,)),
}
#[derive(PartialEq, Debug, Encode, Decode)]
pub enum T {
    One,
}

